File: sandbox/Sandbox.java
// Copyright 2010 owners of the AI Challenge project
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy
// of the License at http://www.apache.org/licenses/LICENSE-2.0 . Unless
// required by applicable law or agreed to in writing, software distributed
// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.
//
// Author: Jeff Cameron (jeff@jpcameron.com)
//
// The sandbox is a wrapper that can be used to invoke untrusted code securely.

import java.io.*;
import java.util.*;

class Sandbox {
    // Sets up the sandbox, but does not invoke it. This constructor sets up a
    // sandbox with no security at all. The spawned process can do whatever it
    // wants. To actually kick off the command, call the Init() method. If
    // trap_sterr is true, then the sandbox catches the stderr output of the
    // spawned program, otherwise the stderr output of the spawned program goes
    // straight to the terminal.
    public Sandbox(String command, boolean trapStderr) {
	childStdin = null;
	childStdout = null;
	childStderr = null;
	this.command = command;
	process = null;
	this.trapStderr = trapStderr;
	isAlive = false;
    }

    // Sets up the sandbox, but does not invoke it. This constructor sets up a
    // sandbox with no security at all. The spawned process can do whatever it
    // wants. To actually kick off the command, call the Init() method. Traps
    // the stderr channel of the spawned process.
    public Sandbox(String command) {
	childStdin = null;
	childStdout = null;
	childStderr = null;
	this.command = command;
	process = null;
	this.trapStderr = true;
	isAlive = false;	
    }

    // The Init() method actually sets up the sandbox and invokes the command
    // inside it. Returns 1 on success, 0 on failure.
    public int Init() {
	try {
	    process = Runtime.getRuntime().exec(command);
	} catch (Exception e) {
	    return 0;
	}
	if (process == null) {
	    return 0;
	}
	childStdin = new OutputStreamWatcher(process.getOutputStream());
	childStdout = new InputStreamWatcher(process.getInputStream());
	childStderr = new InputStreamWatcher(process.getErrorStream());
	childStdin.start();
	childStdout.start();
	childStderr.start();
	return 1;
    }

    // Destroys the spawned process.
    public void Kill() {
	if (process != null) {
	    process.destroy();
	    childStdin.Kill();
	    childStdout.Kill();
	    childStderr.Kill();
	}
	process = null;
    }

    // Returns true if the spawned process is thought to be alive. If something
    // is known to be wrong with the process, then false is returned. Examples
    // include if an IO function has failed or if the Kill() function has been
    // called.
    public boolean IsAlive() {
	return process != null;
    }

    // Writes the given string to the stdin channel of the spawned process,
    // followed by a single newline character (\n). On success, returns the
    // number of bytes written. On error, -1 is returned.
    public int WriteLine(String message) {
	Write(message);
	childStdin.AddCharacters("\n");
	return message.length() + 1;
    }

    // Writes the given string to the stdin channel of the spawned process,
    // without adding a newline character to the end. On success, returns the
    // number of bytes written. On error, -1 is returned.
    public int Write(String message) {
	childStdin.AddCharacters(message);
	return message.length();
    }

    // Attempts to read a line from the stdout channel of the spawned process.
    // This method does not block.
    //
    // If a complete line of the program's output is available, it is placed
    // into the buf variable and the method returns the number of characters
    // read. The terminating newline character is not included in buf, but is
    // included in the returned character count.
    //
    // If a complete line of the program's output is not yet available, null is
    // returned.
    public String ReadLine() {
	String s = childStdout.Readline();
	return s;
    }

    // A blocking ReadLine. Waits for characters written to the spawned program's
    // stdout stream until a newline is read, or max_blocking_time milliseconds
    // have elapsed. Returns the number of characters read.
    public String ReadLine(int maxBlockingTime) {
	long startTime = System.currentTimeMillis();
	while (System.currentTimeMillis() - startTime < maxBlockingTime) {
	    String s = ReadLine();
	    if (s != null) {
		return s;
	    }
	    try {Thread.sleep(10); } catch (Exception e) { }
	}
	return null;
    }

    // Reads a line from the spawned program's stderr channel. Works just the
    // same as the ReadLine method, except for stderr instead of stdout.
    public String ReadErrorLine() {
	return childStderr.Readline();
    }

    // Returns the shell command used to initialize this sandbox.
    public String Command() {
	return command;
    }

    // Continuously monitors an InputStream in a separate thread.
    class InputStreamWatcher extends Thread {
        public InputStreamWatcher(InputStream inputStream) {
	    this.inputStream = inputStream;
	    buffer = "";
	    lines = new LinkedList<String>();
	    dieDieDie = false;
	}

	public void run() {
	    while (!dieDieDie) {
		if (!Iteration()) {
		    return;
		}
		try {Thread.sleep(10); } catch (Exception e) { }
	    }
	}

	// Kills this thread. Stops monitoring the InputStream.
	public synchronized void Kill() {
	    dieDieDie = true;
	}

	// Checks to get one line of input, if available. If no line is
	// immediately available, returns null.
	public synchronized String Readline() {
	    return lines.poll();
	}

	// Checks the stream. Returns false if there's a problem. If everything
	// is peachy, returns true.
	private synchronized boolean Iteration() {
	    if (inputStream == null) {
		return false;
	    }
	    int c = -1;
	    try {
		int avail = inputStream.available();
		if (avail > 0) {
		    c = inputStream.read();
		}
	    } catch (Exception e) {
		return false;
	    }
	    if (c < 0) {
		return true;
	    }
	    if (c == '\n') {
		lines.offer(buffer);
		buffer = new String("");
	    } else {
		buffer += (char)c;
	    }
	    return true;
	}

	private InputStream inputStream;
	private String buffer;
	private Queue<String> lines;
	private boolean dieDieDie;
    }

    // Continuously monitors an OutputStream in a separate thread.
    class OutputStreamWatcher extends Thread {
        public OutputStreamWatcher(OutputStream inputStream) {
	    this.outputStream = outputStream;
	    buffer = "";
	    dieDieDie = false;
	}

	public void run() {
	    while (!dieDieDie) {
		if (!Iteration()) {
		    return;
		}
		try {Thread.sleep(10); } catch (Exception e) { }
	    }
	}

	// Kills this thread. Stops monitoring the InputStream.
	public synchronized void Kill() {
	    dieDieDie = true;
	}

	// Adds to the queue of characters that are being sent to the sandboxed
	// process' stdin channel.
	public synchronized void AddCharacters(String s) {
	    buffer += s;
	}

	// Writes characters from the buffer to the sandboxed process' stdin
	// channel.
	private synchronized boolean Iteration() {
	    if (outputStream == null) {
		return false;
	    }
	    if (buffer.length() == 0) {
		return true;
	    }
	    try {
		outputStream.write(buffer.getBytes());
		outputStream.flush();
		buffer = new String("");
	    } catch (Exception e) {
		return false;
	    }
	    return true;
	}

	private OutputStream outputStream;
	private String buffer;
	private boolean dieDieDie;
    }

    // Three file descriptors that connect to the stdin, stdout, and stderr
    // streams of the spawned process.
    private OutputStreamWatcher childStdin;
    private InputStreamWatcher childStdout;
    private InputStreamWatcher childStderr;

    // The shell command to be invoked inside this sandbox.
    private String command;

    // The spawned process.
    private Process process;

    // Whether or not to trap the spawned program's stderr output.
    private boolean trapStderr;

    // Keeps track of whether the spawned process is in good shape.
    private boolean isAlive;
}

File: sandbox/sandbox.cc
// Copyright 2010 owners of the AI Challenge project
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy
// of the License at http://www.apache.org/licenses/LICENSE-2.0 . Unless
// required by applicable law or agreed to in writing, software distributed
// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.
//
// Author: Jeff Cameron (jeff@jpcameron.com)

#include "sandbox/sandbox.h"
#include <fcntl.h>
#include <iostream>
#include <sched.h>
#include <signal.h>
#include <stdlib.h>
#include <string>
#include <ctime>
#include <vector>
#include "cpp_util/string_util.h"

Sandbox::Sandbox(const std::string& command, bool trap_stderr) {
  child_stdin_ = -1;
  child_stdout_ = -1;
  child_stderr_ = -1;
  command_ = command;
  pid_ = 0;
  trap_stderr_ = trap_stderr;
  is_alive_ = false;
  child_stdout_thread_ = -1;
  child_stderr_thread_ = -1;
  child_stdout_buffer_length_ = 0;
  child_stderr_buffer_length_ = 0;
}

Sandbox::~Sandbox() {
  Kill();
}

int Sandbox::Init() {
  // Tokenize the program arguments.
  std::vector<std::string> tokens;
  StringUtil::Tokenize(command_, " ", tokens);
  char **argv = new char*[tokens.size()+1];
  if (argv == NULL) {
    return 0;
  }
  for (unsigned int i = 0; i < tokens.size(); ++i) {
    argv[i] = (char*)tokens[i].c_str();
  }
  argv[tokens.size()] = NULL;
  // Set up pipes to communicate with the client.
  int child_stdin_pipe[2];
  int child_stdout_pipe[2];
  int child_stderr_pipe[2];
  if (pipe(child_stdin_pipe) ||
      pipe(child_stdout_pipe) ||
      pipe(child_stderr_pipe)) {
    delete argv;
    return 0;
  }
  // Exec the client process.
  pid_ = fork();
  if (pid_ == -1) {
    delete argv;
    return 0;
  } else if (pid_ == 0) {
    dup2(child_stdin_pipe[0], 0);
    dup2(child_stdout_pipe[1], 1);
    if (trap_stderr_) {
      dup2(child_stderr_pipe[1], 2);
    }
    execv(argv[0], argv);
    exit(1); // We should never get here because execv shouldn't return.
  }
  // Set up the pipes to be non-blocking.
  child_stdin_ = child_stdin_pipe[1];
  child_stdout_ = child_stdout_pipe[0];
  if (trap_stderr_) {
    child_stderr_ = child_stderr_pipe[0];
  } else {
    child_stderr_ = -1;
  }
  SetNonBlockingIO(child_stdin_);
  SetNonBlockingIO(child_stdout_);
  if (trap_stderr_) {
    SetNonBlockingIO(child_stderr_);
  }
  delete argv;
  // Start up some threads to continuously monitor the client's stdout and
  // stderr streams and buffer any output in memory.
  child_stdout_thread_ = fork();
  if (child_stdout_thread_ == -1) {
    Kill();
    return 0;
  } else if (child_stdout_thread_ == 0) {
    ChildStdoutMonitor();
  }
  if (trap_stderr_) {
    child_stderr_thread_ = fork();
    if (child_stderr_thread_ == -1) {
      Kill();
      return 0;
    } else if (child_stderr_thread_ == 0) {
      ChildStderrMonitor();
    }
  }
  is_alive_ = true;
  return 1;
}

void Sandbox::Kill() {
  is_alive_ = false;
  if (pid_ == 0) {
    return;
  }
  kill(pid_, SIGTERM);
  kill(child_stdout_thread_, SIGTERM);
  kill(child_stderr_thread_, SIGTERM);
  int child_pid = fork();
  if (child_pid == -1) {
    return;
  } else if (child_pid == 0) {
    sleep(1);
    kill(pid_, SIGKILL);
    kill(child_stdout_thread_, SIGKILL);
    kill(child_stderr_thread_, SIGKILL);
    exit(0);
  }
}

std::string Sandbox::Command() {
  return command_;
}

bool Sandbox::IsAlive() {
  return is_alive_;
}

int Sandbox::SetNonBlockingIO(int fd)
{
  int flags;
#if defined(O_NONBLOCK) // If they have O_NONBLOCK, use the Posix way to do it.
  if (-1 == (flags = fcntl(fd, F_GETFL, 0))) {
    flags = 0;
  }
  return fcntl(fd, F_SETFL, flags | O_NONBLOCK);
#else // Otherwise, use the old way of doing it.
  flags = 1;
  return ioctl(fd, FIOBIO, &flags);
#endif
}

int Sandbox::WriteLine(const std::string& message) {
  std::string with_newline = message + "\n";
  return write(child_stdin_, with_newline.c_str(), with_newline.length());
}

int Sandbox::Write(const std::string& message) {
  return write(child_stdin_, message.c_str(), message.length());
}

void Sandbox::SlideCharactersBack(char *s, int start, int end) {
  for (int i = 0; i < end - start; ++i) {
    s[i] = s[start + i];
  }
}

int Sandbox::ReadLine(std::string& buf, int max_blocking_time) {
  clock_t start_time = clock();
  while (true) {
    int result = ReadLine(buf);
    if (result > 0) {
      return result;
    }
    clock_t current_time = clock();
    int elapsed_ms = (current_time - start_time) * 1000 / CLOCKS_PER_SEC;
    if (elapsed_ms >= max_blocking_time) {
      break;
    }
    sched_yield();
  }
  return 0;
}

int Sandbox::ReadErrorLine(std::string& buf) {
  std::string line;
  for (int i = 0; i < child_stderr_buffer_length_; ++i) {
    char c = child_stderr_buffer_[--child_stderr_buffer_length_];
    if (c == '\n') {
      buf = line;
      SlideCharactersBack(child_stderr_buffer_,
			  i + 1,
			  child_stderr_buffer_length_);
      child_stderr_buffer_length_ -= i;
      return i + 1;
    } else {
      line += c;
    }
  }
  return 0;
}

int Sandbox::getcpid() {
  return pid_;
}

int Sandbox::ReadLine(std::string& buf) {
  std::string line;
  //std::cout << child_stdout_buffer_length_;
  for (int i = 0; i < child_stdout_buffer_length_; ++i) {
    char c = child_stdout_buffer_[--child_stdout_buffer_length_];
    std::cout << "popping character: " << c << std::endl;
    if (c == '\n') {
      buf = line;
      SlideCharactersBack(child_stdout_buffer_,
			  i + 1,
			  child_stdout_buffer_length_);
      child_stdout_buffer_length_ -= i;
      return i + 1;
    } else {
      line += c;
    }
  }
  return 0;
}

void Sandbox::ChildStdoutMonitor() {
  char c;
  while (true) {
    //std::cout << child_stdout_buffer_length_;
    ssize_t bytes_read = read(child_stdout_, &c, 1);
    if (bytes_read > 0) {
      std::cout << "pushing character: " << c << std::endl;
      child_stdout_buffer_[child_stdout_buffer_length_++] = c;
    }
    sched_yield();
  }
}

void Sandbox::ChildStderrMonitor() {
  char c;
  while (true) {
    ssize_t bytes_read = read(child_stderr_, &c, 1);
    if (bytes_read > 0) {
      child_stderr_buffer_[child_stderr_buffer_length_++] = c;
    }
    sched_yield();
  }
}
